File: 312-burst-balloons/burst-balloons.cpp
class Solution {
public:
    int solve(vector<int> &nums, int i, int j, vector<vector<int>> &dp){
        if(i > j) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        if(i == j){
            int temp = nums[i];
            if(i-1 >= 0) temp *= nums[i-1];
            if(j+1 < nums.size()) temp *= nums[j+1];
            return dp[i][j] = temp;
        } else {
            int ans = INT_MIN;
            for(int k=i; k<=j; k++){
                int temp = nums[k];
                if(i-1 >= 0) temp *= nums[i-1];
                if(j+1 < nums.size()) temp *= nums[j+1];

                ans = max(ans, temp + solve(nums,i,k-1,dp) + solve(nums,k+1,j,dp));
            }
            return dp[i][j] = ans;
        }
    }

    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n, -1));
        return solve(nums,0,n-1,dp);
    }
};

File: 312-burst-balloons/burst-balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:

        dp = [[0] * (len(nums)+2) for _ in range(len(nums)+2)] 

        # def recur(i, j):
        #     if i > j:
        #         return 0
            
        #     if dp[i][j] != -1:
        #         return dp[i][j]
            
        #     for k in range(i, j+1):
        #         dp[i][j] = max(dp[i][j], (arr[k]*arr[i-1]*arr[j+1]) + recur(i, k-1) + recur(k+1, j))
            
        #     return dp[i][j]
        
        # arr = [1] + nums + [1]
        # return recur(1, len(nums))

        arr = [1] + nums + [1]
        for i in range(len(nums), 0, -1):
            for j in range(1, len(nums)+1):
                if i > j:
                    continue
                for k in range(i, j+1):
                    dp[i][j] = max(dp[i][j], (arr[k]*arr[i-1]*arr[j+1]) + dp[i][k-1] + dp[k+1][j])
        
        return dp[1][len(nums)]


            
            
